<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="BTCtools : Beyond the consensus NGS tools" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>BTCtools</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/josephhughes/btctools">View on GitHub</a>

          <h1 id="project_title">BTCtools</h1>
          <h2 id="project_tagline">Command-line tutorial</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/josephhughes/btctools/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/josephhughes/btctools/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>Here I will try to explain how you can run the various btctools from the command-line. The core of btctools is written in perl and makes use of the the <a href="http://search.cpan.org/~lds/Bio-SamTools/">Bio-SamTools module</a> written by Lincoln D. Stein. 
        So the various perl script can be run from the command-line to automate the processing on multiple bam files.</p>
        
        <p>There are two basic processing scripts: <a href="#btcutils">btcutils</a> which is used for processing a BAM file and <a href="#btcmerge">btcmerge</a> which is used for combining multiple btc outputs.
     
<h3>
<a name="btcutils" class="anchor" href="#btcutils"><span class="octicon octicon-link"></span></a>btcutils</h3>

<p>This script is used to generate the files that contain the frequencies of various mutations from the bam file. In the most basic form, you may just be interested in the number of SNPs at each site of your alignment. From the directory where the perl script is located:</p>

<pre><code>$ perl btcutils.pl -bam path/to/input.bam -ref path/to/reference.fasta -stub out
</code></pre>

<p>Bear in mind that the bam file needs its associated index file (.bai). All of the arguments above are necessary to run btcutils. The output from the latter command will be 4 files:</p>
<p><ul>
<li>out_entropy.txt which contains the frequency of mutations at each individual site for each of the gene segments</li>
<li>out_log.txt which contains the distribution of mutations per read and the average entropy per gene segment</li>
<li>out_motif.fa a fasta file with 10 bases either side of each mutation</li>
<li>out_read.txt the per read position mutation count and average quality</li>
</ul>
</p>

<p>It is also possible to provide information about the number of non-synonymous and synonymous mutations at each amino acid site, but to do this you need to provide a text file (e.g. CodingRegion.txt) with the information about the protein names and open reading frames, in the following tab delimited format:</p>
<pre><code>
Protein	Beg	End	Reference
PB2	28	2307	PB2_Influenza
PB1	25	2301	PB1_Influenza
PB1-F2	119	391	PB1_Influenza
PA	25	2176	PA_Influenza
PA-X_A	25	596	PA_Influenza
PA-X_B	598	784	PA_Influenza
</code></pre>

<p>The first column corresponds to the protein name, the names in this column must be unique, the second column corresponds to the start of the orf for that protein, the third is the end of the orf for the protein and the fourth corresponds to the identifier (name) of the reference in your reference file from which the protein is transcribed. Run the command:</p>

<pre><code>$ perl btcutils.pl -bam path/to/input.bam -ref path/to/reference.fasta -orfs CodingRegions.txt -stub out
</code></pre>

<p>The latter command produces all of the previous 4 files with an additional out_AA.txt with the information about the non-synonymous, synonymous and stop-codons.</p>

<h3>
<a name="btcmerge" class="anchor" href="#btcmerge"><span class="octicon octicon-link"></span></a>btcmerge</h3>

<p>Once you have produced multiple btc files, you may want to merge several, either to compare the results from independent replicates, or to compare the intra-host diversity from different individuals, or even to compare closely related viruses in the same or different individuals.</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">BTCtools maintained by <a href="https://github.com/josephhughes">josephhughes</a> and <a href="https://github.com/rjorton">rjorton</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
